### PR TITLE
Support `WRANGLER_CI_MATCH_TAG` environment variable

### PR DESCRIPTION
**Pull request by penalosa**
## What this PR solves / how to test

Fixes [DEVX-1399](https://jira.cfdata.org/browse/DEVX-1399)

## Author has addressed the following

- Tests
  - [ ] TODO (before merge)
  - [x] Included
  - [ ] Not necessary because:
- E2E Tests CI Job required? (Use "e2e" label or ask maintainer to run separately)
  - [ ] I don't know
  - [x] Required / Maybe required
  - [ ] Not required because:
- Changeset ([Changeset guidelines](https://github.com/cloudflare/workers-sdk/blob/main/CONTRIBUTING.md#changesets))
  - [ ] TODO (before merge)
  - [x] Included
  - [ ] Not necessary because:
- Public documentation
  - [ ] TODO (before merge)
  - [ ] Cloudflare docs PR(s): 
  - [x] Not necessary because: Internal detail for CI use

### CODE REVIEW

File: packages/wrangler/src/deploy/index.ts [hunk 1]
```diff
@@ -326,6 +328,15 @@ export async function deployHandler(
 
 	const beforeUpload = Date.now();
 	const name = getScriptName(args, config);
+	assert(
```
Comment (CarmenPopoviciu) [inline packages/wrangler/src/deploy/index.ts:331]: what is the diff between using `assert` over `logger.error()` here?

File: packages/wrangler/src/environment-variables/misc-variables.ts [hunk 2]
```diff
@@ -89,3 +89,12 @@ export const getOutputFileDirectoryFromEnv = getEnvironmentVariableFactory({
 export const getOutputFilePathFromEnv = getEnvironmentVariableFactory({
 	variableName: "WRANGLER_OUTPUT_FILE_PATH",
 });
+
+/**
+ * `WRANGLER_CI_MATCH_TAG` specifies a worker tag
+ *
+ * If this is set, Wrangler will ensure the worker being targeted has this tag
```
Comment (CarmenPopoviciu) [inline packages/wrangler/src/environment-variables/misc-variables.ts:96]:
```suggestion
 * If this is set, Wrangler will ensure the Worker being targeted has this tag
```

File: packages/wrangler/src/match-tag.ts [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+		const worker = await fetchResult<ServiceMetadataRes>(
+			`/accounts/${accountId}/workers/services/${workerName}`
+		);
+		tag = worker.default_environment.script.tag;
+	} catch (e) {
+		logger.debug(e);
+		// code: 10090, message: workers.api.error.service_not_found
+		if ((e as { code?: number }).code === 10090) {
+			throw new FatalError(
+				`Your Worker's name (${workerName}) does not match what is expected by the CI system`
```
Comment (CarmenPopoviciu) [inline packages/wrangler/src/match-tag.ts:30]: is there an action we can recommend to users here?

File: packages/wrangler/src/match-tag.ts [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+			throw new FatalError(
+				"Wrangler cannot validate that your Worker name matches what is expected by the CI system"
+			);
+		}
+	}
+	if (tag !== matchTag) {
+		logger.debug(
+			`Failed to match Worker tag. The API returned "${tag}", but the CI system expected "${matchTag}"`
+		);
+		throw new FatalError(
```
Comment (CarmenPopoviciu) [inline packages/wrangler/src/match-tag.ts:42]: is it worth merging these two error messages into one, and surfacing everything in the `FatalError`?

Comment (CarmenPopoviciu) [reply packages/wrangler/src/match-tag.ts:42]: IMHO I shouldn't be in debug mode to get that kind of error details

File: packages/wrangler/src/versions/index.ts [hunk 5]
```diff
@@ -244,7 +246,19 @@ export async function versionsUploadHandler(
 	const accountId = args.dryRun ? undefined : await requireAuth(config);
 	const name = getScriptName(args, config);
 
-	await standardPricingWarning(config);
+	assert(
```
Comment (CarmenPopoviciu) [inline packages/wrangler/src/versions/index.ts:249]: same `assert` vs `logger.error` question here

File: packages/wrangler/src/versions/index.ts [hunk 6]
```diff
@@ -244,7 +246,19 @@ export async function versionsUploadHandler(
...
+		name,
+		'You need to provide a name when publishing a worker. Either pass it as a cli arg with `--name <name>` or in your config file as `name = "<name>"`'
+	);
+
+	if (!args.dryRun) {
```
Comment (CarmenPopoviciu) [inline packages/wrangler/src/versions/index.ts:254]: it look like the two `if` blocks can be merged into one?

File: packages/wrangler/src/versions/index.ts [hunk 7]
```diff
@@ -244,7 +246,19 @@ export async function versionsUploadHandler(
 	const accountId = args.dryRun ? undefined : await requireAuth(config);
 	const name = getScriptName(args, config);
 
-	await standardPricingWarning(config);
+	assert(
+		name,
+		'You need to provide a name when publishing a worker. Either pass it as a cli arg with `--name <name>` or in your config file as `name = "<name>"`'
```
Comment (CarmenPopoviciu) [inline packages/wrangler/src/versions/index.ts:251]:
```suggestion
		'You need to provide a name when publishing a Worker. Either pass it as a cli arg with `--name <name>` or in your config file as `name = "<name>"`'
```

**Review by CarmenPopoviciu — APPROVED** [review]

File: packages/wrangler/src/match-tag.ts (hunk 3 above)
Comment (jahands) [reply packages/wrangler/src/match-tag.ts:30]: We do have some error messaging written up [here](https://jira.cfdata.org/browse/BANDA-411). cc @aninibread

**Pull request merged**